File: android/src/main/java/com/reactnativenotificationbanner/Appearance.kt
package com.reactnativenotificationbanner

import android.graphics.Bitmap
import android.graphics.Color

fun argb(alpha: Float, red: Float, green: Float, blue: Float): Int {
  return (alpha * 255.0f + 0.5f).toInt() shl 24 or
    ((red * 255.0f + 0.5f).toInt() shl 16) or
    ((green * 255.0f + 0.5f).toInt() shl 8) or
    (blue * 255.0f + 0.5f).toInt()
}

@Suppress("EnumEntryName")
enum class BannerStyle {
  error, info, success
}

interface Style {
  var icon: Bitmap?
  var backgroundColor: Int
  var titleColor: Int
  var messageColor: Int

  var titleFont: Font
}

class Font(
  val size: Double,
  val family: String?,
  var textAlign: String
)

class ErrorStyle(
  override var icon: Bitmap?,
  override var backgroundColor: Int,
  override var titleColor: Int,
  override var messageColor: Int,
  override var titleFont: Font
) : Style

class SuccessStyle(
  override var icon: Bitmap?,
  override var backgroundColor: Int,
  override var titleColor: Int,
  override var messageColor: Int,
  override var titleFont: Font
) : Style

class InfoStyle(
  override var icon: Bitmap?,
  override var backgroundColor: Int,
  override var titleColor: Int,
  override var messageColor: Int,
  override var titleFont: Font
) : Style

class DefaultStyle(
  var elevation: Float,
  var cornerRadius: Float,
  var margin: Int,
  var padding: Float,
  var minWidth: Int,
  var maxWidth: Int,


  var success: SuccessStyle,
  var error: ErrorStyle,
  var info: InfoStyle
)

var currentAppearance = DefaultStyle(
  elevation = 4f,
  cornerRadius = 1000f,
  margin = 32,
  padding = 16f,
  minWidth = 150,
  maxWidth = 300,
  success = SuccessStyle(
    backgroundColor = argb(1f, 0.22f, 0.8f, 0.46f),
    titleColor = Color.WHITE,
    messageColor = Color.WHITE,
    titleFont = Font(size = 17.0, family = null, textAlign = "center"),
    icon = null
  ),

  error = ErrorStyle(
    backgroundColor = argb(1f, 0.9f, 0.31f, 0.26f),
    titleColor = Color.WHITE,
    messageColor = Color.WHITE,
    titleFont = Font(size = 17.0, family = null, textAlign = "center"),
    icon = null
  ),

  info = InfoStyle(
    backgroundColor = argb(1f, 0.23f, 0.6f, 0.85f),
    titleColor = Color.WHITE,
    messageColor = Color.WHITE,
    titleFont = Font(size = 17.0, family = null, textAlign = "center"),
    icon = null
  )
)

fun bannerAppearance(type: BannerStyle): Style {
  return when (type) {
    BannerStyle.error -> currentAppearance.error
    BannerStyle.info -> currentAppearance.info
    BannerStyle.success -> currentAppearance.success
  }
}

fun styleFrom(type: String): BannerStyle {
  return when (type) {
    "error" -> BannerStyle.error
    "info" -> BannerStyle.info
    else -> BannerStyle.success
  }
}

fun setNewAppearance(appearance: DefaultStyle) {
  currentAppearance = appearance
}

File: ios/NotificationBanner/Appearance.swift
//
//  Appearance.swift
//  react-native-notification-banner
//
//  Created by Sergei Golishnikov on 06/11/2021.
//

import Foundation


protocol Style {
    var icon: UIImage? { get }
    var backgroundColor: UIColor { get }
    var titleColor: UIColor { get }
    var messageColor: UIColor { get }
    
    var titleFont: Font { get }
}

struct Shadow {
    let offset: CGSize
    let color: UIColor
    let radius: CGFloat
    let opacity: Float
}

struct Font: Codable {
    var size: Double
    var family: String?
    var textAlign: String
}

struct ErrorStyle: Style {
    var icon: UIImage?
    var backgroundColor: UIColor
    var titleColor: UIColor
    var messageColor: UIColor
    var titleFont: Font
}

struct SuccessStyle: Style {
    var icon: UIImage?
    var backgroundColor: UIColor
    var titleColor: UIColor
    var messageColor: UIColor
    var titleFont: Font
}

struct InfoStyle: Style {
    var icon: UIImage?
    var backgroundColor: UIColor
    var titleColor: UIColor
    var messageColor: UIColor
    var titleFont: Font
}

struct DefaultStyle {
    var cornerRadius: CGFloat = 0
    var margin: CGFloat = 32
    var padding: CGFloat = 16
    var maxWidth: CGFloat = UIScreen.main.bounds.width - 32
    var minWidth: CGFloat = 150
    
    var shadow: Shadow
    
    var success: SuccessStyle
    var error: ErrorStyle
    var info: InfoStyle
}

var currentAppearance = DefaultStyle(
    cornerRadius: 1000,
    margin: 32,
    padding: 16,
    shadow: Shadow(offset: .init(width: 3, height: 1), color: .black, radius: 10, opacity: 0.5),
    success: SuccessStyle(
        icon: nil,
        backgroundColor: UIColor(red: 0.22, green: 0.80, blue: 0.46, alpha: 1.00),
        titleColor: .white,
        messageColor: .white,
        titleFont: Font(size: 17, family: nil, textAlign: "center")
    ),
    error: ErrorStyle(
        icon: nil,
        backgroundColor: UIColor(red: 0.90, green: 0.31, blue: 0.26, alpha: 1.00),
        titleColor: .white,
        messageColor: .white,
        titleFont: Font(size: 17, family: nil, textAlign: "center")
    ),
    info: InfoStyle(
        icon: nil,
        backgroundColor: UIColor(red: 0.23, green: 0.60, blue: 0.85, alpha: 1.00),
        titleColor: .white,
        messageColor: .white,
        titleFont: Font(size: 17, family: nil, textAlign: "center")
    )
)

func titleFont(type: BannerStyle) -> UIFont {
    let style = bannerAppearance(type: type)
    var font = UIFont.systemFont(ofSize: style.titleFont.size, weight: .bold)
    if let family = style.titleFont.family {
        font = UIFont(name: family, size: font.pointSize)!
    }
    return font
}

func messageFont(type: BannerStyle) -> UIFont {
    let style = bannerAppearance(type: type)
    return .systemFont(ofSize: 15)
}

func bannerAppearance(type: BannerStyle) -> Style {
    switch type {
    case .danger:
        return currentAppearance.error
    case .info:
        return currentAppearance.info
    case .success:
        return currentAppearance.success
    case .warning:
        return currentAppearance.error
    case .customView:
        fatalError("unsupported")
    }
}

func setNewAppearance(_ appearance: DefaultStyle) {
    currentAppearance = appearance
}


